### ISSUE TITLE
RFC - Horovod as a Research Platform: Gradient Compression, Sparsification and various distributed strategies

### ISSUE THREAD

**Issue by DEKHTIARJonathan** [horovod/horovod#1157]
### Context

After some talks at ICML and NeurIPS during one of the workshops we co-organize with @wbailer, we figured out that Horovod would potentially be a very interesting framework to facilitate research on optimizing distributed DL workloads. 

Aforementioned Workshops:
- ICML: https://icml.cc/Conferences/2019/ScheduleMultitrack?event=3520
- NeurIPS: https://nips.cc/Conferences/2018/Schedule?showEvent=10941

### Objective

We would like to provide researchers a workflow on distributed training optimization strategies. This should come as a platform to develop the proposed method and a way to evaluate the efficiency of a given method.

By providing a unified package accross DL frameworks, Horovod is the only framework which address the issue of "not everyone use the same framework". And at the same time, the core design of the Horovod library makes it very easy to integrate inside Horovod.

### Proposition

We would like to open the discussion on how Horovod could be used to facilitate and benchmark distributed training optimization strategies like the following:

- **Gradient Compression:** Quantization, Low Precision, etc.
- **Gradient Sparsification:** Block Sparse, other Sparse Data Formats, Sparsification strategies
- **Reduced Communication Strategies:** Distributed Parallel SGD, Stochastic Gradient Push, etc.
- Other ?

It shall be possible to design a Python/C++/CUDA package providing a standard API compliant with Horovod which makes it virtually plug'n'play in the library. This idea is inspired by the followings in Tensorflow:
- https://github.com/tensorflow/custom-op
- https://www.tensorflow.org/guide/extend/op

This package shall be as simple as possible to use allowing researchers to focus on the core of their method. All of them may not know how MPI/NCCL works if they focus on Gradient Quantization for example.

### How would Horovod benefits from the research ?

Aside of the additional exposure, it would be very interesting for Horovod to have a large number of methods almost plug'n'play with the library. The most promising methods could even be merged upstream in `horovod.contrib` or `horovod.research` or provided through an additional package `horovod_extensions`

----------------

As a final word, please let us know (myself and @wbailer) if you have any question. 
I tried to keep it short and to the point.

Looking forward hearing from you,

Best,

Jonathan

**Comment by tgaddair** [comment]
Hey @DEKHTIARJonathan, thanks for drafting this!  Having a unified "plugin" API for different algorithms is something we've mused about off and on in the past, but haven't yet put together anything concrete.  

Most of our past discussions on the topic have been around gradient compression (in addition to the Deep Gradient Compression you mentioned, there's also [QSGD-TF](https://github.com/ganshaoduo/QSGD-TF), which forks Horovod and would ideally be integrated as a plugin or similar).  Not long ago we introduced the [Compression API](https://github.com/horovod/horovod/blob/master/horovod/tensorflow/compression.py) for FP16, with the intent of adding more methods that conform to this interface, but so far we haven't integrated anything at the C++ layer.

A lot of the work we've been doing recently has been in refactoring the [operations.cc](https://github.com/horovod/horovod/blob/master/horovod/common/operations.cc) monolith into framework-level components with dynamic operation selection at runtime (see: [operation_manager.cc](https://github.com/horovod/horovod/blob/master/horovod/common/ops/operation_manager.cc)). Part of the goal of this undertaking was to make it easy for contributors to drop in new frameworks, algorithms, etc., and should hopefully make it easier to build the kind of plugin API you're proposing.

Looking forward to having some more conversations about this and what you all have in mind for next steps!

**Comment by DEKHTIARJonathan** [comment]
Wow I never heard of this implementation: https://github.com/ganshaoduo/QSGD-TF ! Very nice work. I will probably test it in the coming weeks.

I know about the Compression API, TBH it's this changes which gave me this idea of using HVD as research platform.

Let me know how we can help you to converge to this point and how you would like the discussion to be conducted ;)

**Comment by armandmcqueen** [comment]
+1 for the idea. I've been considering whether it makes sense to implement other distribution strategies (stochastic gradient push in particular) inside of Horovod

**Comment by DEKHTIARJonathan** [comment]
@armandmcqueen If you wanna go in this direction. Please get me in the loop. I'd love to help.
Personal Email: contact@jonathandekhtiar.eu

Nonetheless, I still believe we need some "standardized" API to integrate with HVD and working on a "per-method" basis would be probably too labor intensive. Today, this paper from Facebook "stochastic gradient push" looks promising, tomorrow it's another one. My primary wish and goal is to make HVD the perfect research platform to develop and integrate any new idea. This would allow easy integration in all DL frameworks and easy benchmarking and comparison. All of this can later be used and integrated upstream or in an extension package for the most interesting methods therefore profiting to every HVD users ;)

Eventually

**Comment by armandmcqueen** [comment]
For the immediate future, I'll be investigating the [PyTorch implementation of SGP](https://github.com/facebookresearch/stochastic_gradient_push), but I'd be very happy to see a plugin API that makes it easier to bring these techniques to Horovod.

**Comment by alsrgv** [comment]
@DEKHTIARJonathan, @armandmcqueen, thanks for starting this discussion!

We're definitely excited to see all the research that uses Horovod to push the boundaries of distributed deep learning. It is our desire to make extending Horovod as simple as possible.

Refactorings that @tgaddair mentioned are a big step in this direction - today, it's much easier to define custom op implementation or compression scheme than it was before, but surely there are more areas to tackle.

Do you have any specific ideas in mind in terms of API extensibility?

Separately, I think it'd be a great idea to integrate Horovod into more DL benchmarks.  It is integrated into [TensorFlow benchmarks](https://github.com/horovod/horovod/blob/master/docs/benchmarks.rst), but surely there's more ground to cover, too.

**Comment by DEKHTIARJonathan** [comment]
>> Do you have any specific ideas in mind in terms of API extensibility?

On top of my head, I see 3 importants behaviors.

- Being able to modify the gradient before send
- Being able to modify the gradients received
- Modify the way Horovod all reduce the gradient, in other term the communication strategy.

1 and 2 are probably quite simple and can be integrated in the form of a callback and standardised. You send gradients you receive a modify version that's it.

The 3rd one is considerably more challenging and TBH I don't have any idea on how it could be achieved in a generic way.

How do you enable switching from a ring all-reduce to any other all reduce topology to a completely other strategy is the biggest challenge IMHO here. Probably having a look to the Pytorch repo on SGP by Facebook could give you a good overview of what kind oglf behaviours researchers would like MODIFY. Without restraining the discussion to this paper it's probably a good example for this scenario.

**Comment by armandmcqueen** [comment]
I don't have any specific ideas on how to design a plugin system, but one current barrier for me is the lack of a developer guide. The docs are great about explaining what Horovod/ring-allreduce are and how to use it, but I'm not aware of any documentation explaining the C++ codebase. If I want to implement SGP I expect to have to do a fair amount of thinking and work, but right now I don't even know what part of the Horovod codebase to start reading.

Going further, some simple tutorials that show how to add a new all-reduce algorithm or compression scheme (trivial but illustrative examples) could be very helpful in teaching users how to get started and what files will probably need to be worked on.

**Comment by DEKHTIARJonathan** [comment]
@tgaddair @alsrgv any update ?
I would like to help and I'm not sure where to start and how kickstart the more technical aspects of the discussion.

**Comment by alsrgv** [comment]
@armandmcqueen, you've raised very good points!  @tgaddair, are we close to putting out dev guide that you were working on?

@DEKHTIARJonathan, I think points 1 and 2 in your list are covered by compression algorithms, e.g. https://github.com/horovod/horovod/blob/master/horovod/tensorflow/compression.py.  And the third point should be addressed by the dev guide.

**Comment by tgaddair** [comment]
I can make it a point to get the dev guide out later this week, and I'll add sections covering adding custom operations / compression algorithms.

**Comment by romerojosh** [comment]
To add a bit to this discussion, are there any plans to have some official support/procedure for adding custom GPU kernels to Horovod written in CUDA? The unmerged PR #453 contained several of these and I can see this being useful in the context of enabling more experimentation with Horovod.

**Comment by alsrgv** [comment]
@romerojosh, I think we can support those by calling nvcc compiler. We have an idea to move all Horovod compilation to CMake, which should make it easier.

#453 was not merged because of convergence issues, but would be good to pick up if somebody's willing!

**Comment by romerojosh** [comment]
@alsrgv Yes, exposing some compile option via CMake to compile in CUDA code would be excellent! The mechanism that #453 used was a little hacky (overrides the compiler used in `setup.py`). Getting a documented/official means to do this from you guys would be great.

**Comment by tgaddair** [comment]
Here are the docs covering our current process for adding custom compression / operations to Horovod (pre send, post receive, and communication strategy): https://github.com/horovod/horovod/blob/master/docs/contributors.rst

Hopefully we can build upon this to make the process even simpler for researchers.

**Comment by DEKHTIARJonathan** [comment]
Thanks a lot @tgaddair. So I had some fun with your docs today. And the first feedback I can give you is the following:
- Very simple and easy to use/build.
- Does not support Compression to SparseTensor => Quite a problem IMHO
- Not that simple to give argument to the compressor. We should provide a "standard way" to give hyper parameters to Compressors.

I give you the most stupid and straightforward example to illustrate:
```
class SparseCompressor(Compressor):
    sparse_threshold = 1e-3

    """Compress all gradients to a sparse gradient"""
    @staticmethod
    def compress(tensor):
        """Downcasts the tensor to 16-bit."""
        tensor_compressed = tensor
        # return tensor_compressed, tensor_compressed.dtype

        if tensor.dtype.is_floating:
            idx = tf.where(tf.math.greater(tensor, SparseCompressor.sparse_threshold))
            tensor_compressed = tf.SparseTensor(idx, tf.gather_nd(tensor, idx), tensor.get_shape())
        return tensor_compressed, tensor_compressed.dtype

    @staticmethod
    def decompress(tensor, ctx):
        """Upcasts the tensor to the initialization dtype."""
        tensor_decompressed = tensor

        print(type(tensor))

        if isinstance(tensor, tf.SparseTensor):
            tensor_decompressed = tf.sparse_tensor_to_dense(tensor)

        return tensor_decompressed
```

**Comment by weberxie** [comment]
Hi, how is this going? Looking forward to this great feature!

**Comment by bryandeng** [comment]
Any updates?

PyTorch may be planning something similar:
https://github.com/pytorch/pytorch/issues/39272
